### DISCUSSION TITLE
Layouts for the different operating systems

### DISCUSSION THREAD

**Discussion by Lobo-Feroz** [binarybottle/engram-es#34] [General]
Here I'll be uploading the keyboard layouts for the different operating systems.

**Windows**

Keyboard Layout Creator 1.4 source file:
[engram-es-v1-20211103.klc.zip](https://github.com/binarybottle/engram-es/files/7478521/engram-es-v1-20211103.klc.zip)
Updated 2021-11-03:
* Added ordinals (ª, º) on the AltGr layer of 'a' and 'o'

**Mac**

[Engram-ESv1.zip](https://github.com/binarybottle/engram-es/files/7066560/Engram-ESv1.zip) by @NickG13 

**Linux**

Next one in the pipeline

**Comment by Lobo-Feroz** [comment]
Engram-es for windows.

* On MS KLC 1.4 the name of the keyboard has to be 8 characters or less, "engram-es" has 9 chars, so I set it as "Engrama" (spanish word for "Engram")

Here are the main layers - base, shift, altgr, altgr+shift, accute accent:

![engram-es - base](https://user-images.githubusercontent.com/10242383/131014318-e92c843f-0fcf-4967-a74b-daeb7cb1451e.png)

![engram-es - shift](https://user-images.githubusercontent.com/10242383/131014347-281f13cf-7607-49b2-b7ce-e2647aa83029.png)

![engram-es - altgr](https://user-images.githubusercontent.com/10242383/131014359-fb3dbdd7-740d-4ab7-9a48-d293257b19e7.png)

![engram-es - altgr - shift](https://user-images.githubusercontent.com/10242383/131014370-21acf4c5-d68f-401a-a465-08abebcbd6e6.png)

![engram-es - accute accent](https://user-images.githubusercontent.com/10242383/131014888-6e94a232-7c49-48e0-9926-231875e438d7.png)

**Comment by NickG13** [comment]
I'm going to make the Mac version then! Ukelele sometimes gives bugs, so hopefully we can do it!
TBH maybe Keyman is a more universal option but I don't understand that software.

**Reply by NickG13** [reply]
Here is the Mac version
[Engram-ESv1.zip](https://github.com/binarybottle/engram-es/files/7066560/Engram-ESv1.zip)

**Reply by binarybottle** [reply]
Keyman is pretty easy once you get the hang of it. I just don't know what its limitations are with respect to diacritical keys.

**Comment by binarybottle** [comment]
@NickG13 @Lobo-Feroz -- The engram repo has a "keymaps" folder that contains linux and windows subfolders for installing on those OSes (if not using Keyman).  Shall I create a similar folder for this repo, and what files should I include in it?  Tha above for Win & Mac?  Do you have installation instructions for people to follow?  Anybody want to give Keyman a whirl?  I hear it's much easier on Windows through their gui than by hand, which is the way I've been doing it.

**Reply by Lobo-Feroz** [reply]
Hey @binarybottle,

Sorry, got swamped in real life. My intended roadmap was this:

- linux layout (in ubuntu)
- keyman (I decided to tackle this system after reading your good opinion on it)
- monkeytype
- announce the keyboard on forums (like reddit) to get more people to test it

Any setup of the repository that you consider useful, I'm all for it. If it mirrors the current engram-en, the better.

A place to show also the typing websites that support engram-es would be nice too. I made a discussion for it, but it could go in some kind of webpage somewhere. Type-fu is done, once the keyboard has been tested by more people and it's final, the author wants to add it to the type-fu list of keyboards.

And by the way, I'm already using engram-es in my keyboard. Have even switched the keypad. It will be a long journey until I reach the speed of decades of experience with qwerty, but it feels much much better in the fingers and the wrists.

**Reply by binarybottle** [reply]
That's a fantastic road map!

What do you mean by "setup of the repository"?  Do you mean any data independent of the corpus and word lists (which should exist some place like OSF or Zenodo)?

What do you think about having the Spanish version of the engram, with links to all of the stuff that you're doing, on the engram.dev website?

**Reply by Lobo-Feroz** [reply]
Hey @binarybottle , sorry for not replying to this either.

> That's a fantastic road map!

Thanks! Linux XKB already functional with limitations, I'm starting keyman next.

> What do you mean by "setup of the repository"? Do you mean any data independent of the corpus and word lists (which should exist some place like OSF or Zenodo)?

Whatever you set up for the placement of the files, documentation and so on, will be OK. For instance:

> What do you think about having the Spanish version of the engram, with links to all of the stuff that you're doing, on the engram.dev website?

Yes, that would be awesome of course :) . Whatever you need from me just tell me, I'm a bit more busy these days but I'll keep hacking at stuff wherever I can.

I'll publish a page in my website as well, but focusing more on the benefits on ergonomics and productivity for the end users, leaving out the most technical details. And it will be in the spanish language.

**Reply by binarybottle** [reply]
I don't know how to set up multiple pages / subdomains in github-pages, so I added a link at the top of engram.dev that points to the engram-es repo.  I also copied the engram-es README.md to README-en.md -- please feel free to translate all or part of this README.md into Spanish, and submit a PR!  Also, I copied the engram repo's keylayouts/{linux,windowi}/README.md files to this repo, so please feel free to modify and add all relevant files by PR!

**Comment by Lobo-Feroz** [comment]
Here's the XKB symbols file for the linux layout:

[engram-es.xkb-symbols.2021-09-11.txt](https://github.com/binarybottle/engram-es/files/7147553/engram-es.xkb-symbols.2021-09-11.txt)

Sorry for the long delay @binarybottle. I had to learn how XKB works, and it's very under-the-hood and barebones, to say the least. For instance, if the file is broken it won't tell you about any errors. Naive me thought that the days of vi-editing system config files were already behind me.

Furthermore, it's far from finished. Somehow the "default" symbols for layers 3 and 4 keep appearing when not overwritten by this file (see image), I need to resolve this. Also you can't set ´ + n = ñ in the xkb file, because it's defined in the locale (it prints ń). You need to edit other files to create a custom dead key. @iandoug mentioned this in earlier posts and now I understand what you meant.

@iandoug: could I ask you some XKB questions in a new thread? or would it be better if I go ask directly in stackexchange and similar places?

But all in all it's been an interesting experience seein how X kb layouts work. Also, the linux layout has a lot of interesting symbols in the alternative layers, like AltGr + Shift + C = ©. It has symbols for the asturian dotted H and L that @NickG13 once mentioned to me: Ḥ, Ḷ. Greek letters like µ. And all sorts of strange stuff. Maybe some of these could be included down the road in engram, or maybe not, since really their usefulness is limited, but it's curious to see how linux layouts are fulled to the brim with symbols in all layers.

Here's the layout as seen in the keyboard manager:

![engam-es-linux-2021-09-10](https://user-images.githubusercontent.com/10242383/132943986-e6651e7c-27b1-4b43-85d8-0b13b2d71228.png)

**Reply by iandoug** [reply]
> Furthermore, it's far from finished. Somehow the "default" symbols for layers 3 and 4 keep appearing when not overwritten by this file (see image), I need to resolve this. 

They are probably defined in
include "latin(type4)"

>Also you can't set ´ + n = ñ in the xkb file, because it's defined in the locale (it prints ń). You need to edit other files to create a custom dead key. @iandoug mentioned this in earlier posts and now I understand what you meant.

This would be handled by the separate .XCompose file. You define "deadkeys" in xkb, then it seems that .XCompose  has instructions like 
<dead_acute> <eng>                      : "ŋ́"   # LATIN SMALL LETTER ENG WITH ACUTE
or in your case
<dead_acute> c                      : "ç"   # LATIN SMALL LETTER C WITH cidilla
<dead_acute> n                      : "ñ"   # LATIN SMALL LETTER N WITH tilda

 /usr/share/X11/locale/en_US.UTF-8/Compose has the defaults, but users create own versions like
/home/ian/.XCompose

```
 $ head /home/ian/.XCompose
# Import default rules from the system Compose file
include "%L"
<Multi_key> <p> <i>             : "π"   U03C0 # Greek small letter pi
<Multi_key> <p> <h>             : "φ"   U03C6 # Greek small letter phi
<Multi_key> <r> <r>             : "ρ"   U03C1 # Greek small letter rho
<Multi_key> <p> <r>             : "Ⳡ"   U2CE1 # COPTIC CAPITAL LETTER OLD NUBIAN NYI .... for φρ
```

> 
> @iandoug: could I ask you some XKB questions in a new thread? or would it be better if I go ask directly in stackexchange and similar places?

Ask :-)

Will try and wrap up my corpus stuff this weekend.

Note: last I heard, Gnome handles XCompose differently to the other desktops, in a not-nice way. This directly affects default Ubuntu versions.

Cheers, Ian

**Reply by iandoug** [reply]
> . Also, the linux layout has a lot of interesting symbols in the alternative layers, like AltGr + Shift + C = ©

See https://eurkey.steffen.bruentjen.eu/

Compose key offers hundreds of other predefined glyphs, or you can make your own.

**Reply by Lobo-Feroz** [reply]
Thanks for the help @iandoug. Let me revise it and get a new version without the latin(type4) and I'll open a new discussion about the rest of issues.

**Reply by binarybottle** [reply]
Thank you both!

**Comment by binarybottle** [comment]
> ![imagen](https://user-images.githubusercontent.com/10242383/131022945-dff359d2-03e9-447a-8111-399e94b4e9db.png)

@Lobo-Feroz and @iandoug -- I am including a description of engram-es in a revision of the engram manuscript under review and would really like to include a high-res version of the layout.  Would either of you have a better image?

**Reply by iandoug** [reply]
How do we differentiate between * on 0 and magic * ...?
Or can we use a different symbol?

The diacritic circle thingies tend to be problematic, depending on font.  Will see what I can do.

**Reply by iandoug** [reply]
Diacritics problematic.
![engram-es-kle](https://user-images.githubusercontent.com/12673886/134957596-4487f482-9265-4975-9235-0da827024b64.jpg)
![engram-es](https://user-images.githubusercontent.com/12673886/134957600-21d065be-a834-49f9-bfc1-277c139fac58.png)
![engram-es](https://user-images.githubusercontent.com/12673886/134957605-b9c04f69-0492-4720-b315-18329144e4d9.jpg)

**Reply by iandoug** [reply]
green tinge is screen artefact.... like Moire pattern.

**Reply by binarybottle** [reply]
Thank you, @iandoug!

**Reply by Lobo-Feroz** [reply]
> 
> 
> > ![imagen](https://user-images.githubusercontent.com/10242383/131022945-dff359d2-03e9-447a-8111-399e94b4e9db.png)
> 
> @Lobo-Feroz and @iandoug -- I am including a description of engram-es in a revision of the engram manuscript under review and would really like to include a high-res version of the layout. Would either of you have a better image?

That's a screen capture of the type-fu website. I've zoomed in and grabbed a bigger screenshot:

![engram-es-typefu-hr](https://user-images.githubusercontent.com/10242383/134961736-fe579ee5-67fa-4269-a1b0-bb7767ad3a3b.png)

I don't know if I can get it at a higher resolution, being from a website and all. @iandoug 's layouts look very cool as well !

Notice: here I've added the ordinals º and ª in the altgr layer of O and A. They were missing in the previous version.

**Reply by binarybottle** [reply]
Thank you, @Lobo-Feroz!

**Reply by iandoug** [reply]
Forgot the magic star. Pity about the diacritics on the 3D version ... maybe you can photoshop?
![engram-es](https://user-images.githubusercontent.com/12673886/134965404-76f9c4d2-19b6-4a22-be62-17d3b3f9d162.png)
![engram-es-kle](https://user-images.githubusercontent.com/12673886/134965410-4e6226db-0928-4163-bc51-b84811a1b087.jpg)

**Comment by binarybottle** [comment]
@iandoug -- your version has a cedilla placed on the C key as an altgr-generated symbol, but that is accessed by the diacritical key instead, so should not be shown.

@iandoug & @Lobo-Feroz -- Please remind me what the star key in the lower left is for, as I thought the diacritical key in the middle of the right center column is supposed to help generate all diacriticals, including those for N and C, correct?

**Reply by iandoug** [reply]
Sorry old eyes ... was NOT expecting a cent symbol ....

![engram-es-kle](https://user-images.githubusercontent.com/12673886/134976015-daee8649-1ea6-4ff3-b162-a54617110ed5.jpg)

**Reply by binarybottle** [reply]
Ha -- thanks!  You generated another, sharper image earlier that was of higher contrast and didn't break up when you zoom in. Would it be possible for you to generate another one like that that is updated?

**Reply by iandoug** [reply]
The 3D version?

The last one is here:
http://www.keyboard-layout-editor.com/#/gists/2ae50a2500bdd3f37b9c5314eb3ab4ad

I zoomed in on the browser to 155% before taking screen shot. The "Export JPG" version messes up the diacritics

The 3d version is here:
https://kle-render.herokuapp.com/

Paste the first URL above into the first field then click Render. Then File -> Save as to save the PNG.
But that version also screws up the diacritics... the font used there is to blame.

**Reply by binarybottle** [reply]
Perfect, THANK YOU!

**Reply by Lobo-Feroz** [reply]
> @iandoug & @Lobo-Feroz -- Please remind me what the star key in the lower left is for, as I thought the diacritical key in the middle of the right center column is supposed to help generate all diacriticals, including those for N and C, correct?

We talked about placing there the asterisk, not a star. The rationale was that the asterisk is a fairly common symbol, so can be placed in that key. While at the same time being redundant, since it's also on the shift+0 key, thus not being a problem with ANSI keyboards not having that key.

Also, the other layers for that key are still unspecified. We have a little more room there, for instance shifting that key is an easy move, so we can place something useful in shift+*

**Reply by binarybottle** [reply]
That's right, thank you very much!

**Reply by iandoug** [reply]
There was discussion earlier about a "magic diacritic key", Arno used the for that, the idea being
 ```   * + n = ñ
   * + c = ç
   * + vowel = vowel acute 
   *  + n = ñ 
    ```
    
Is that now replaced by diacritic key in centre?

**Reply by Lobo-Feroz** [reply]
> Is that now replaced by diacritic key in centre?

Yep, the original ☆ key for diacritics it's basically that. No need for an actual magic key with complex behavior since the normal dead key behavior seems enough (spanish only uses acute accents and the ü). And it has to be in a fairly easy position since the accented vowels are actually quite frequent. This was discussed in the punctuation thread iirc. 

So, the lower left * is just an asterisk :)

**Comment by Lobo-Feroz** [comment]
Hello there!

@NickG13 informed me that a windows user reported that the ordinals (ª, º) were missing in the AltGr layer.

I've updated the .klc file in the main post. Here's a screenshot of the AltGr layer:

![EngramaAltGr](https://user-images.githubusercontent.com/10242383/140421590-b01e6ae9-8f3f-4305-bd02-9106617caea0.jpg)
